File: web/src/core/stage/HtmlCleaningStage.ts
import { PipelineStage } from '../pipeline/PipelineStage.js';
import { ProcessingContext } from '../pipeline/ProcessingContext.js';

/**
 * Stage 2 (cleaning): HTML -> plain text, then strip email quotes, signatures
 * and whitespace noise. Runs on every source text before block splitting.
 *
 * The Java version used JSoup; here we stay dependency-free so the same code
 * runs in the browser, in Node and in tests. Block-level tags become newlines
 * so list structure survives for the acceptance-criteria stage.
 */
const HTML_HINT = /<\s*(p|div|br|span|table|ul|ol|li|b|i|strong|em|h[1-6]|a)\b/i;
const EMAIL_QUOTE_HEADER = /^On .{5,120} wrote:\s*$/im;
const QUOTED_LINE = /^\s*>.*$/gm;
const SIGNATURE = /^--\s*$[\s\S]*/m;
const EXCESS_BLANK_LINES = /\n{3,}/g;
const TRAILING_SPACES = /[ \t]+$/gm;

const NAMED_ENTITIES: Record<string, string> = {
  '&nbsp;': ' ',
  '&lt;': '<',
  '&gt;': '>',
  '&quot;': '"',
  '&#39;': "'",
  '&apos;': "'",
  '&amp;': '&',
};

function decodeEntities(text: string): string {
  let out = text;
  for (const [entity, value] of Object.entries(NAMED_ENTITIES)) {
    // &amp; is decoded last (it is last in insertion order) to avoid double-decoding.
    out = out.split(entity).join(value);
  }
  out = out.replace(/&#(\d+);/g, (_, n: string) => String.fromCodePoint(parseInt(n, 10)));
  out = out.replace(/&#x([0-9a-fA-F]+);/g, (_, n: string) => String.fromCodePoint(parseInt(n, 16)));
  return out;
}

function htmlToText(html: string): string {
  let t = html;
  t = t.replace(/<\s*br\s*\/?\s*>/gi, '\n'); // line breaks
  t = t.replace(/<\s*li[^>]*>/gi, '- '); // list items get a bullet
  // Block boundaries become newlines. Note: <ul>/<ol> themselves do NOT emit a
  // newline, so a heading stays in the same block as the list it introduces
  // (the acceptance-criteria section extractor depends on that).
  t = t.replace(/<\/?\s*(p|div|li|tr|h[1-6])[^>]*>/gi, '\n');
  t = t.replace(/<[^>]+>/g, ''); // strip remaining inline tags
  t = decodeEntities(t);
  return t;
}

export function clean(raw: string): string {
  let text = raw;
  if (HTML_HINT.test(text)) {
    text = htmlToText(text);
  }
  // Drop everything from an email-quote header onward.
  const quoteHeader = EMAIL_QUOTE_HEADER.exec(text);
  if (quoteHeader) {
    text = text.substring(0, quoteHeader.index);
  }
  text = text.replace(QUOTED_LINE, '');
  text = text.replace(SIGNATURE, '');
  text = text.replace(TRAILING_SPACES, '');
  text = text.replace(EXCESS_BLANK_LINES, '\n\n');
  return text.trim();
}

export class HtmlCleaningStage implements PipelineStage {
  readonly name = 'html-cleaning';

  process(context: ProcessingContext): void {
    for (const source of context.sources) {
      source.text = clean(source.text);
    }
  }
}

File: crunchy-core/src/main/java/io/crunchy/core/stage/HtmlCleaningStage.java
package io.crunchy.core.stage;

import io.crunchy.core.pipeline.PipelineStage;
import io.crunchy.core.pipeline.ProcessingContext;
import io.crunchy.core.pipeline.SourceText;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.util.regex.Pattern;

/**
 * Stage 2 (cleaning): HTML → plain text, then strip email quotes, signatures
 * and whitespace noise. Runs on every source text before block splitting.
 */
public final class HtmlCleaningStage implements PipelineStage {

    private static final Pattern HTML_HINT = Pattern.compile("<\\s*(p|div|br|span|table|ul|ol|li|b|i|strong|em|h[1-6]|a)\\b", Pattern.CASE_INSENSITIVE);
    private static final Pattern EMAIL_QUOTE_HEADER = Pattern.compile("(?im)^On .{5,120} wrote:\\s*$");
    private static final Pattern QUOTED_LINE = Pattern.compile("(?m)^\\s*>.*$");
    private static final Pattern SIGNATURE = Pattern.compile("(?s)(?m)^--\\s*$.*");
    private static final Pattern EXCESS_BLANK_LINES = Pattern.compile("\\n{3,}");
    private static final Pattern TRAILING_SPACES = Pattern.compile("(?m)[ \\t]+$");

    @Override
    public String name() {
        return "html-cleaning";
    }

    @Override
    public void process(ProcessingContext context) {
        for (SourceText source : context.getSources()) {
            source.setText(clean(source.getText()));
        }
    }

    static String clean(String raw) {
        String text = raw;
        if (HTML_HINT.matcher(text).find()) {
            // Preserve line structure: mark block boundaries with a sentinel that
            // survives Jsoup's whitespace normalization, then restore newlines.
            Document doc = Jsoup.parse(text);
            doc.select("br").after("\\n");
            // Wrapping block tags on both sides puts a blank line between
            // adjacent paragraphs, so <p>A</p><p>B</p> stays two blocks.
            doc.select("p, div, tr, h1, h2, h3, h4, h5, h6").prepend("\\n").append("\\n");
            // List items only close with a newline (never open with one), so a
            // heading stays in the same block as the list it introduces — the
            // acceptance-criteria section extractor depends on that. <ul>/<ol>
            // deliberately contribute no break for the same reason.
            doc.select("li").prepend("- ").append("\\n");
            text = doc.text().replace("\\n", "\n");
        }
        // Drop everything from an email-quote header onward.
        var quoteHeader = EMAIL_QUOTE_HEADER.matcher(text);
        if (quoteHeader.find()) {
            text = text.substring(0, quoteHeader.start());
        }
        text = QUOTED_LINE.matcher(text).replaceAll("");
        text = SIGNATURE.matcher(text).replaceAll("");
        text = TRAILING_SPACES.matcher(text).replaceAll("");
        text = EXCESS_BLANK_LINES.matcher(text).replaceAll("\n\n");
        return text.strip();
    }
}
